### DISCUSSION TITLE
Any way to add a template without overriding?

### DISCUSSION THREAD

**Discussion by aljawaid** [aljawaid/KanboardEmailHistory#12] [Q&A] [answered]
https://github.com/aljawaid/KanboardEmailHistory/blob/009d1ff9b65dbd7b740be79c087b9b6251c7dffc/Plugin.php#L17

@creecros I use this plugin, so I guess it's overriding the standard templates... but can I not use another copy of the file just for this plugin?  I can't figure out how to do it without an override, as the template is already overriden by Autosubtasks, and my version is specific to this plugin as it has a placeholder which other plugins may not need.

```php
public function initialize()
	{
		$this->actionManager->register(new EmailTaskHistory($this->container));
		
		if (!file_exists('plugins/AutoSubtasks')) {
		    $this->template->setTemplateOverride('action_creation/params', 'kanboardEmailHistory:action_creation/params');
		}
    		
	}
```

**Comment by aljawaid** [comment, marked as answer]
#11

**Reply by creecros** [reply]
you can't. you would have to put the logic in autosubtasks to account for your plugin.

or, add hooks in kb core code, then update both plugins to use hooks.

**Reply by aljawaid** [reply]
could this work?
```
public function initialize()
	{
		$this->actionManager->register(new EmailTaskHistory($this->container));
		
		if (!file_exists('plugins/AutoSubtasks')) {
		    $this->template->setTemplateOverride('action_creation/params', 'kanboardEmailHistory:action_creation/params');
		} else {
			$this->template->setTemplateOverride('autoSubtasks:action_creation/params', 'kanboardEmailHistory:action_creation/params');
    		
	}
```

then I could make all my changes in my own version

**Reply by aljawaid** [reply]
Need a way around this somehow

**Reply by creecros** [reply]
I gave you your 2 options, there are no more.

you have 2 plugins, overriding the same template. nothing you do in this plugin, will stop the other plugin from doing what it does.

**Reply by aljawaid** [reply]
> I gave you your 2 options, there are no more.
> 
> you have 2 plugins, overriding the same template. nothing you do in this plugin, will stop the other plugin from doing what it does.

I understand, I'm just trying to find a way to work it out.

About the core, what if I override the `step 2`?

autosubtasks overrides the `params.php` but what if i override step 2, which is `action_creation/event.php` ? then in that override, we say, if this plugin then process and go to MY params file, otherwise leave it as it is.... this will then be picked up by autosubtasks and overridden.

```php
<div class="page-header">
    <h2><?= t('Choose an event') ?></h2>
</div>

// SO IF WE DO AN `IF` STATEMENT HERE WITH MAYBE A CUSTOM CONTROLLER...
IF ...EXISTS PLUGIN/KANBOARDEMAILHOSTORY
     then duse my controller to send my version of the params file, otherwise....
ELSE... BELOW IS DEFAULT CODE

<form method="post" action="<?= $this->url->href('ActionCreationController', 'params', array('project_id' => $project['id'])) ?>">
    <?= $this->form->csrf() ?>

    <?= $this->form->hidden('action_name', $values) ?>

    <?= $this->form->label(t('Action'), 'action_name') ?>
    <?= $this->form->select('action_name', $available_actions, $values, array(), array('disabled')) ?>

    <?= $this->form->label(t('Event'), 'event_name') ?>
    <?= $this->form->select('event_name', $events, $values) ?>

    <div class="form-help">
        <?= t('When the selected event occurs execute the corresponding action.') ?>
    </div>

    <?= $this->modal->submitButtons(array(
        'submitLabel' => t('Next step')
    )) ?>
</form>
```

hope youre not angry with me

**Reply by creecros** [reply]
it's good thinking, to be honest. a bit extra work, but definitely possible.

**Reply by aljawaid** [reply]
> it's good thinking, to be honest. a bit extra work, but definitely possible.

excellent... we found a way... I will create an issue for it.... definitely feel free to join in lol

**Reply by aljawaid** [reply]
Override params.php Without Affecting Other Plugins #15
